File: 1.two-sum.js
/*
 * @lc app=leetcode id=1 lang=javascript
 *
 * [1] Two Sum
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

// create hashtable to store
// key: value of nums
// value: index
// if exists key: (target - nums[i]) then you got the result

var twoSum = function (nums, target) {
    var hashTable = {};
    for (var i = 0; i < nums.length; i++) {
        // if exist key 
        if (hashTable[nums[i]]) {
            hashTable[nums[i]].push(i)
        }
        else {
            hashTable[nums[i]] = [i]
        }

        // for exp: (9-2) = 7
        // but 7 existed in hashtable before => 
        // return result: index of 7 and current index
        let prev = hashTable[target - nums[i]]
        if (prev && prev[0] != i) {
            return [prev[0], i]
        }
    }
    return []
};
// @lc code=end

console.log(twoSum([3,2,4], 6));

File: 1.two-sum.py
#
# @lc app=leetcode id=1 lang=python
#
# [1] Two Sum
#

# solution 2: using Hash map
# using hashmap to store all indicates previous index
# [value] = index
# result: existed (target - value[i]) && that value != i 

# @lc code=start
class Solution(object):
    def twoSum(self, nums, target):
        my_dict = {}
        for index, value in enumerate(nums):
            if(my_dict.get(value) is None):
                my_dict[value] = index
            if (my_dict.get(target - value) is not None and my_dict.get(target - value) != index):
                return [my_dict.get(target - value), index]
        
        return []
        
# @lc code=end

# ✅ Đoạn để debug - viết test case ở đây
if __name__ == "__main__":
    sol = Solution()
    result = sol.twoSum([2,7,11,15], 9)
    print("Result:", result)